### ISSUE TITLE
Cannot pickle

### ISSUE THREAD

**Issue by crazyplum** [cdgriffith/Box#36]
```
pickle.dump(Box({}), open('/tmp/test.p', 'wb'))
test = pickle.load(open('/tmp/test.p', 'rb'))
```

caused endless `KeyError: '_box_config'` and then `RecursionError: maximum recursion depth exceeded while calling a Python object`

python version: 3.6

**Comment by cdgriffith** [comment]
Thanks for the catch @crazyplum! (Until fix, workaround would be to simply 'to_dict' and `from_dict' before / after pickling). 

In dev branch, the recursion issue is fixed, but the underlying problem is the missing `_box_config` key when `pickle` tries to unpickle it.  I will have to dig into pickling to fix this https://docs.python.org/3/library/pickle.html#pickling-class-instances

**Comment by cdgriffith** [comment]
This is fixed in a working branch https://github.com/cdgriffith/Box/tree/conversion_box_raise_errors

Due to difference in pickle protocols, had to both add  `__setstate__` and  `__new__` methods that work around the special way `_box_config` works.

**Comment by cdgriffith** [comment]
Fixed in Added in 3.1 https://github.com/cdgriffith/Box/commit/252e2f574213708cbbcd42bdec442434384127fe
